src/components/network/network-loader.js

```diff
@@ -19,117 +19,286 @@ genotet.utils.inherit(genotet.NetworkLoader, genotet.ViewLoader);
-  var regex;
-  try {
-    regex = RegExp(geneRegex, 'i');
-  } catch (e) {
-    genotet.error('invalid gene regex', geneRegex);
-    return;
-  }
-  this.data.nodes = _.filter(this.data.nodes, function(node) {
-    return !node.id.match(regex);
+genotet.NetworkLoader.prototype.addGene_ = function(genes) {
+  var oldGenes = {};
+  this.data.network.nodes.forEach(function(node) {
+    oldGenes[node.id] = true;
   });
-  this.data.edges = _.filter(this.data.edges, function(edge) {
-    return !edge.source.match(regex) && !edge.target.match(regex);
+  var newGenes = [];
+  genes.forEach(function(gene) {
+    if (!(gene in oldGenes)) {
+      newGenes.push(gene);
+    }
   });
 
-  this.updateGeneRegex_();
-  this.signal('geneRemove');
+  if (newGenes.length == 0) {
+    genotet.warning('Genes are already in graph');
+    return;
+  }
+
+  var params = {
+    type: 'add-gene',
+    fileName: this.data.network.fileName,
+    gene: newGenes,
+    nodes: this.data.network.nodes
+  };
+
+  this.get(genotet.data.serverURL, params, function(data) {
+    var isTF = {};
+    this.data.networkInfo.nodes.forEach(function(node) {
+      isTF[node.id] = node.isTF;
+    });
+    newGenes.forEach(function(gene) {
+      this.data.network.nodes.push({
+        id: gene,
+        label: gene,
+        isTF: isTF[gene]
+      });
+    }.bind(this));
+    data.edges.forEach(function(edge) {
+      this.data.network.edges.push(edge);
+    }.bind(this));
+  }.bind(this), 'cannot add genes');
 };
 
 /**
- * Updates the regex that selects the current gene set. Because of gene
- * removal, unlike gene addition, it is hard to incrementally modify the regex.
- * This function sets the gene regex to a verbose concatenation of the gene
- * ids. This is only called in removeGenes_().
+ * Delete one gene from graph and related edges.
+ * @param {!Array<string>} genes Names of the genes.
  * @private
  */
-genotet.NetworkLoader.prototype.updateGeneRegex_ = function() {
-  var regex = '';
-  this.data.nodes.forEach(function(node, index) {
-    regex += node.id + (index == this.data.nodes.length - 1 ? '' : '|');
-  }, this);
-  this.data.geneRegex = regex;
+genotet.NetworkLoader.prototype.deleteGene_ = function(genes) {
+  var geneMap = {};
+  genes.forEach(function(gene) {
+    geneMap[gene] = true;
+  });
+  // delete the nodes
+  for (var i = 0; i < this.data.network.nodes.length; i++) {
+    if (this.data.network.nodes[i].id in geneMap) {
+      this.data.network.nodes.splice(i, 1);
+    }
+  }
+  // delete the edges
+  for (var i = this.data.network.edges.length - 1; i >= 0; i--) {
+    if (this.data.network.edges[i].source in geneMap ||
+      this.data.network.edges[i].target in geneMap) {
+      this.data.network.edges.splice(i, 1);
+    }
+  }
 };
 
 /**
- * Fetches the incident edges of a given node.
- * @param {!Object} node Node of which the incident edges are queried.
+ * Adds one edge to the graph.
+ * @param {!Object} data Input data for adding an edge.
```

Comment:
Be specific on Object.

```
@param {{
  network: ...
  networkInfo: ...
}}
```
But in this case it could be clearer to just use two function arguments.